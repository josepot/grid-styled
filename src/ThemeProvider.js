import {Component, Children} from 'react'
import PropTypes from 'prop-types'

export default class ThemeProvider extends Component {
  constructor(props, ctx) {
    super(props, ctx);
  }

  getChildContext() {
    return {theme: this.props.theme}
  }

  render() {
    return Children.only(this.props.children)
  }
}

export const themeShape = {
  breakpoints: PropTypes.arrayOf(PropTypes.number),
  space: PropTypes.arrayOf(PropTypes.number),
  fontSizes: PropTypes.arrayOf(PropTypes.number),
};

ThemeProvider.propTypes = {
  theme: PropTypes.shape(themeShape).isRequired,
  children: PropTypes.element.isRequired,
}

ThemeProvider.childContextTypes = {
  theme: PropTypes.shape(themeShape),
}
