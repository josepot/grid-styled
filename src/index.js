export { default as Box } from './Box'
export { default as Grid } from './Grid'
export { default as Flex } from './Flex'
export { default as Half } from './Half'
export { default as Third } from './Third'
export { default as Quarter } from './Quarter'
export { default as Golden } from './Golden'
export { default as ThemeProvider } from './ThemeProvider'
export { default as withBox } from './withBox'
export { default as withFlex } from './withFlex';
