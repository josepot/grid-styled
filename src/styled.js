import {styled} from 'styletron-react';

export default (component, ...styles) =>
  styled(component, props =>
    styles
      .map(style => (typeof style === 'function' ? style(props) : style))
      .filter(x => !!x)
      .reduce((res, style) => Object.assign(res, style), {})
  )
