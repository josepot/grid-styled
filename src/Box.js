import React from 'react'
import { space, width, removeProps, propTypes } from 'styled-system'
import styled from './styled'

const CleanDiv = cleanElement('div');

CleanDiv.propTypes = {
  ...propTypes.width
}

export const flex = ({flex}) => flex ? {flex} : null

const Box = styled(CleanDiv,
  {boxSizing: 'border-box'},
  width,
  space,
  flex
)




export default Box
