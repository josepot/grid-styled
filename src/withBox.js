import React from 'react'
import { space, responsiveStyle, width } from 'styled-system'
import { number, string, array, oneOfType } from 'prop-types'
import tag from './tag-hoc'
import propTypes from './propTypes'
import removeProps from './remove-props'
import styled from './styled'

const flex = responsiveStyle('flex')
const order = responsiveStyle('order')

const cache = {};
const Tag = tag(removeProps)

export default base => {
  if (cache[base]) return cache[base];

  const Base = Tag(base)

  const Box = styled(Base,
    { boxSizing: 'border-box' },
    width,
    space,
    flex,
    order
  )
  Box.displayName = 'Box'

  const responsivePropType = oneOfType([
    number,
    string,
    array
  ])

  Box.propTypes = Object.assign({}, propTypes, {
    flex: responsivePropType,
    order: responsivePropType
  })

  if (typeof base === 'string') cache[base] = Box;

  return Box;
}
