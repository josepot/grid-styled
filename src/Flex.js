import React from 'react'
import Box from './Box'
import styled from './styled'

const align = ({align}) => align ? {alignItems: align} : null
const justify = ({justify}) => justify ? {justifyContent: justify} : null
const order = ({order}) => order ? {order} : null
const wrap = ({wrap}) => wrap ? {flexWrap: 'wrap'} : null
const column = ({column}) => column ? {flexDirection: 'column'} : null

const Flex = styled(Box,
  {display: 'flex'},
  align,
  justify,
  order,
  wrap,
  column
)

export default Flex
