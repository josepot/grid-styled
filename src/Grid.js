import Box from './Box'
import styled from './styled'

const align = props => ({ verticalAlign: props.align || 'top' })

const Grid = styled(Box, {display: 'inline-block'}, align)

export default Grid
