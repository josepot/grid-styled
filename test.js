import test from 'ava'
import React from 'react'
import {create} from 'react-test-renderer'
import Styletron from 'styletron-server'
import {StyletronProvider} from 'styletron-react';
import {
  Box,
  Grid,
  Flex
} from './src'
import { flex } from './src/Box'

const render = element => {
  const styletron = new Styletron()
  const renderedElement = create(
    <StyletronProvider styletron={styletron}>
    {element}
    </StyletronProvider>
  )
  const styles = styletron.getStylesheetsHtml();
  return {renderedElement, styles};
}


// Box
test('Box renders', t => {
  const box = render(<Box m={2} px={3} />);
  t.snapshot(box)
})

test('Box renders with props', t => {
  const box = render(<Box
    m={[ 1, 2 ]}
    px={[ 1, 2 ]}
    w={1}
    flex='1 1 auto'
  />)
  t.snapshot(box)
})

test('flex util returns null', t => {
  const sx = flex({})
  t.is(sx, null)
})

test('flex util returns a style object', t => {
  const sx = flex({ flex: 'none' })
  t.is(sx.flex, 'none')
})

// Grid
test('Grid renders', t => {
  const grid = render(<Grid />)
  t.snapshot(grid)
})

test('Grid has a classname', t => {
  const {renderedElement} = render(<Grid />)
  t.truthy(renderedElement.toJSON().props.className)
})

// Flex
test('Flex renders', t => {
  const flex = render(<Flex />)
  t.snapshot(flex)
})

