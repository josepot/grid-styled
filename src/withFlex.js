import React from 'react'
import { responsiveStyle } from 'styled-system'
import { oneOfType, number, string, array, bool } from 'prop-types'
import commonPropTypes from './propTypes'
import styled from './styled'
import withBox from './withBox';

const wrap = responsiveStyle('flex-wrap', 'wrap', 'wrap')
const direction = responsiveStyle('flex-direction', 'direction')
const align = responsiveStyle('align-items', 'align')
const justify = responsiveStyle('justify-content', 'justify')
const column = props => props.column ? {flexDirection: 'column'} : null

const responsivePropType = oneOfType([
  number,
  string,
  array,
  bool
])
const propTypes = Object.assign({}, commonPropTypes, {
  wrap: responsivePropType,
  direction: responsivePropType,
  align: responsivePropType,
  justify: responsivePropType,
  column: bool
})
const cache = {}

export default base => {
  if (cache[base]) return cache[base];

  const Flex = styled(withBox(base),
    { display: 'flex' },
    wrap,
    column,
    direction,
    align,
    justify,
  )
  Flex.displayName = 'Flex'
  Flex.propTypes = propTypes

  if (typeof base === 'string') cache[base] = Flex

  return Flex
}
