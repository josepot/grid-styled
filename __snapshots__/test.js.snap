// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders 1`] = `
Object {
  "renderedElement": <div
    className="a b c d e"
/>,
  "styles": "<style class=\\"_styletron_hydrate_\\">.a{-moz-box-sizing:border-box}.b{box-sizing:border-box}.c{margin:16px}.d{padding-left:32px}.e{padding-right:32px}</style>",
}
`;

exports[`Box renders with props 1`] = `
Object {
  "renderedElement": <div
    className="a b c d e f g h i j"
/>,
  "styles": "<style class=\\"_styletron_hydrate_\\">.a{-moz-box-sizing:border-box}.b{box-sizing:border-box}.c{margin:8px}.f{padding-left:8px}.g{padding-right:8px}.h{-webkit-flex:1 1 auto}.i{-ms-flex:1 1 auto}.j{flex:1 1 auto}</style><style class=\\"_styletron_hydrate_\\" media=\\"screen and (min-width: 40em)\\">.d{padding-left:16px}.e{padding-right:16px}</style>",
}
`;

exports[`Flex renders 1`] = `
Object {
  "renderedElement": <div
    className="a b c d e f g"
/>,
  "styles": "<style class=\\"_styletron_hydrate_\\">.a{display:-webkit-box}.b{display:-moz-box}.c{display:-ms-flexbox}.d{display:-webkit-flex}.e{display:flex}.f{-moz-box-sizing:border-box}.g{box-sizing:border-box}</style>",
}
`;

exports[`Grid renders 1`] = `
Object {
  "renderedElement": <div
    className="a b c d"
/>,
  "styles": "<style class=\\"_styletron_hydrate_\\">.a{display:inline-block}.b{vertical-align:top}.c{-moz-box-sizing:border-box}.d{box-sizing:border-box}</style>",
}
`;
